pkg/txbuild/builder.go

```diff
@@ -19,25 +25,80 @@ var (
 	ErrUnsupportedNetworkType = errors.New("unsupported network type")
 )
 
-type MultiBuilder struct {
-	chains data.ChainsQ
+type Builder interface {
+	BuildTx(ctx context.Context, req *resources.BuildTx, rawTxData interface{}) (*resources.UnsubmittedTx, error)
+}
 
-	evmBuild    *ethtx.Builder
-	solanaBuild *soltx.Builder
-	nearBuild   *neartx.Builder
+type MultiBuilder struct {
+	chains   data.ChainsQ
+	builders map[string]Builder
 }
 
 func NewMultiBuilder(cfg config.Config) *MultiBuilder {
-	abi, err := bridge.BridgeMetaData.GetAbi()
+	abi, err := gobind.BridgeMetaData.GetAbi()
 	if err != nil {
 		panic(errors.Wrap(err, "failed to get bridge abi"))
 	}
 
+	networks, err := cfg.Core().TokenManager().GetParams(context.Background())
+	if err != nil {
+		panic(errors.Wrap(err, "failed to get networks"))
+	}
+
+	builders := make(map[string]Builder)
+	contracts := make(map[string]bridgeConfig)
+
+	for _, net := range networks.Networks {
+		bridgeParams := net.GetBridgeParams()
+		if bridgeParams == nil {
+			continue
+		}
+
+		contracts[net.Name] = bridgeConfig{
+			contract:       bridgeParams.Contract,
+			adminPublicKey: bridgeParams.Admin,
+		}
+	}
+
+	for _, chain := range cfg.ChainsQ().List() {
+		chainConf := contracts[chain.Name]
+		if chainConf.contract == "" {
+			panic(fmt.Errorf("contract address not found for chain %s", chain.Name))
+		}
+
+		switch chain.Type {
+		case tokenmanager.NetworkType_EVM:
+			cli, err := ethclient.Dial(chain.Rpc)
+			if err != nil {
+				panic(errors.Wrap(err, "failed to dial eth client"))
+			}
+
+			builders[chain.Name] = ethtx.NewBuilder(cli, abi, common.HexToAddress(chainConf.contract))
```

Comment:
Are these builder structs immutable? Please make sure there are no race conditions when working with this `builders` map when using a `MultiBuilder`

Reply:
MultiBuilder actually creates just once during injecting it into the api router middleware